feat: free fees for thor holders before tip 014

NeOMakinG opened the pull request:
## Description

This is implementing free fees for thor holders before TIP 014 proposal

I don't have enough funds to do a full swap to test if fees are effectively removed, but assuming setting fees to 0 inside `calculateFees` effectively change it

I didn't make the votePower slide systemic on purpose as this thing is supposed to stay for a few month and the lift would be a bit more risky, happy to discuss it in a follow up if it bring value

<!-- Please describe your changes -->

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes #8019

## Risk
Medium, fee calculation could lead user wrongly
> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing
- Check regular trade fees without an account holding THOR, under 1000 and over 1000
- Check trade using `0xC85feF7A1b039A9e080aadf80FF6f1536DADa088` as a spoofed address, to see that trade fees are free because of THOR holding before the proposal
- You could also verify that `new Date().getFullYear() <= 2025` is true and `new Date().getFullYear() <= 2024` is false to verify if it will be disabled programmatically
- Also verify the mixpanel event is sent on trade (if you have enough funds) 
<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

### Engineering
n/a
<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations
n/a
- [x] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)
### Normal account not holding thor > 1000$ trade
<img width="528" alt="image" src="https://github.com/user-attachments/assets/1085efd8-fbdb-428e-b96c-deb5d30ae5b1">

### Normal account < 1000$ trade
<img width="514" alt="image" src="https://github.com/user-attachments/assets/122c53df-147c-4276-8371-5bedfa7cae9b">

### thor holding account > 1000$ trade
<img width="505" alt="image" src="https://github.com/user-attachments/assets/b4e33909-30f7-4319-bba1-afd81ad2634f">

### thor holding account < 1000$ trade
<img width="515" alt="image" src="https://github.com/user-attachments/assets/720b8faa-7993-40a1-a989-377e16324432">

### Banner
<img width="533" alt="image" src="https://github.com/user-attachments/assets/e1b6ee7f-49a8-432b-9510-1fe8cfe7596f">

NeOMakinG commented on src/lib/fees/model.ts:
```diff
@@ -48,15 +53,26 @@ export const calculateFees: CalculateFeeBps = ({ tradeAmountUsd, foxHeld, feeMod
+  const isFree =
+    tradeAmountUsd.lt(noFeeThresholdUsd) ||
+    (isThorFreeEnabled && thorHeld?.isGreaterThanOrEqualTo(THORSWAP_UNIT_THRESHOLD))
+
+  const isThorFree =
+    isThorFreeEnabled &&
+    thorHeld?.isGreaterThanOrEqualTo(THORSWAP_UNIT_THRESHOLD) &&
+    new Date().getFullYear() < THORSWAP_MAXIMUM_YEAR_TRESHOLD
```
Meaty logic: if thorVotingPower > 1 and current year is < 2025

NeOMakinG commented on src/lib/fees/parameters/index.ts:
```diff
@@ -5,14 +5,18 @@ import type { FeeCurveParameters, ParameterModel } from './types'
   SWAPPER: swapperParameters,
   THORCHAIN_LP: thorchainLpParameters,
+  // THOR asset doesn't have any fee models for now as it's not using curves
+  THORSWAP: {} as FeeCurveParameters,
```
We don't need this for now as this swapper doesn't involve any curve

NeOMakinG replied:
We are not consuming this at all, its just to make TS happy

But I liked the zeroed out solution and implemented it in [25dc736](https://github.com/shapeshift/web/pull/8044/commits/25dc7365114443a641a36bd748e76afdffa92ae8) with a little comment

NeOMakinG commented on src/lib/fees/parameters/types.ts:
```diff
@@ -8,4 +8,4 @@ export type FeeCurveParameters = {
 }
 
-export type ParameterModel = 'SWAPPER' | 'THORCHAIN_LP'
+export type ParameterModel = 'SWAPPER' | 'THORCHAIN_LP' | 'THORSWAP'
```
It's not 100% what we want as THORSWAP is not a model, it's a little shortcut, we could make this more systemic and avoid this but the value is not so big for now as it's supposed to last 2 months, wdyt?

NeOMakinG commented on src/state/apis/snapshot/snapshot.ts:
```diff
@@ -94,6 +106,36 @@ export const snapshotApi = createApi({
+      queryFn: async (_, { dispatch }) => {
+        const query = `
+          query {
+            space(id: "${THORSWAP_SNAPSHOT_SPACE}") {
```
We could pass this as a parameter and change the logic, but pretty nitpick from my pov as we will want to kick this out in a few month

NeOMakinG commented on src/state/apis/snapshot/snapshot.ts:
```diff
@@ -152,6 +194,63 @@ export const snapshotApi = createApi({
+          }
+          const evmAddresses = Array.from(
+            accountIds.reduce<Set<string>>((acc, accountId) => {
+              const { account, chainId } = fromAccountId(accountId)
+              isEvmChainId(chainId) && acc.add(account)
+              return acc
+            }, new Set()),
+          )
+          const delegation = false // don't let people delegate for discounts - ambiguous in spec
+          const votingPowerResults = await Promise.all(
+            evmAddresses.map(async address => {
+              const votingPowerUnvalidated = await getVotingPower(
+                address,
+                '1',
+                strategies,
+                THOR_TIP_014_BLOCK_NUMBER,
+                THORSWAP_SNAPSHOT_SPACE,
+                delegation,
+              )
+              // vp is THOR in crypto balance
+              return bnOrZero(VotingPowerSchema.parse(votingPowerUnvalidated).vp)
+            }),
+          )
+          const thorHeld = BigNumber.sum(...votingPowerResults).toNumber()
+
+          // Return an error tuple in case of an invalid foxHeld value so we don't cache an errored value
+          if (isNaN(thorHeld)) {
+            const data = 'NaN thorHeld value'
+            return { error: { data, status: 400 } }
+          }
+
+          dispatch(snapshot.actions.setThorVotingPower(thorHeld.toString()))
+          return { data: thorHeld.toString() }
+        } catch (e) {
+          console.error(e)
+
+          return { error: { data: e, status: 400 } }
+        }
+      },
+    }),
```
Pretty same logic than FOX, but I wanted to keep this sane and separated to avoid implementing regression on the FOX side while this is a quickwin for a few months

0xean commented:
<img width="526" alt="Screenshot 2024-10-30 at 1 48 18 PM" src="https://github.com/user-attachments/assets/34ff615c-a9e0-4154-8572-ad17d5900ce3">


Copy here is unfortunately wrong.  Trades under 1k are already free

"Should say Thor holders trade for free through 2024."

NeOMakinG commented:
> <img alt="Screenshot 2024-10-30 at 1 48 18 PM" width="526" src="https://private-user-images.githubusercontent.com/15096737/381672507-34ff615c-a9e0-4154-8572-ad17d5900ce3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAzMTE0ODcsIm5iZiI6MTczMDMxMTE4NywicGF0aCI6Ii8xNTA5NjczNy8zODE2NzI1MDctMzRmZjYxNWMtYTllMC00MTU0LTg1NzItYWQxN2Q1OTAwY2UzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMwVDE3NTk0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5YzFiOTY5YTVhNzExNGJlNjliODVlMWE5MTU3MDE5NmMzZmY4MGI3OWQzYzdjNTUwY2ZkZGQ4YTgxMjA5MTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RdVKHfb-DqfZWesocTjEUKHu01dAdAw3ajCIHdLoRSQ">
> Copy here is unfortunately wrong. Trades under 1k are already free
> 
> Should say "Thor holders trade for free through 2024"

Updated in [8ea8869](https://github.com/shapeshift/web/pull/8044/commits/8ea88694e7a5cf502e8f2d9da5ef3dd82cc329fd)

<img width="533" alt="image" src="https://github.com/user-attachments/assets/62b901fa-686c-4f70-b4e3-8d3f2dd7bb26">

Also removed the link so we don't block the release with the missing link

woodenfurniture commented on src/state/apis/snapshot/snapshot.ts:
```diff
@@ -18,13 +18,18 @@ import { ProposalSchema, SnapshotSchema, VotingPowerSchema } from './validators'
 const SNAPSHOT_SPACE = 'shapeshiftdao.eth'
+const THORSWAP_SNAPSHOT_SPACE = 'thorswapcommunity.eth'
+
+const THOR_TIP_014_BLOCK_NUMBER = 21072340
```
q: Can we add a link to the snapshot as a code comment?

NeOMakinG replied:
Added in [3969eb7](https://github.com/shapeshift/web/pull/8044/commits/3969eb73a27ecfb4073f3820a983c494efe2c204)

woodenfurniture commented on src/components/MultiHopTrade/components/SharedTradeInput/SharedTradeInput.tsx:
```diff
@@ -44,23 +45,26 @@ export const SharedTradeInput: React.FC<SharedTradeInputProps> = ({
-        </Card>
+      <Center width='inherit' alignItems='flex-start'>
+        <Box width='full' maxWidth='500px'>
+          <ThorFreeFeeBanner />
```
preferably blocking: it should be possible to nest this under the `tradeInputRef` so the height calcs include the height of the banner to avoid the odd-looking gap at the bottom:


<img src="https://github.com/user-attachments/assets/885810fb-3ab7-4c96-afc0-2b3b78743ec9" width="200">


Alternatively we should move the quote list to align with the top of the form excluding the banner to keep the layout looking tidy

NeOMakinG replied:
Aligned with the top of the form as a quick win 
<img width="1096" alt="image" src="https://github.com/user-attachments/assets/ca60dcd8-6b9b-4bf3-9c50-58018bfeecb8">

in [30b7cb9](https://github.com/shapeshift/web/pull/8044/commits/30b7cb926fe202f7e1e9e44fd5a3616439f2f4f9)

woodenfurniture replied:
*** SO *** much better

woodenfurniture commented on src/components/ThorFreeFeeBanner/ThorFreeFeeBanner.tsx:
```diff
@@ -0,0 +1,52 @@
+
+import icons from './asset-icons.png'
+
+// const cardHover = {
+//   textDecoration: 'none',
+//   opacity: '.7',
+// }
```
nitpick:

```suggestion

// const cardHover = {
//   textDecoration: 'none',
//   opacity: '.7',
// }
```

NeOMakinG replied:
I wanted to keep all those comments so we don't forget this stuff when the link is ready to be sent to prod, wdyt?

woodenfurniture replied:
Ah got it, can add a todo if it's less distracting if you want

NeOMakinG replied:
Added in [68096ca](https://github.com/shapeshift/web/pull/8044/commits/68096ca8f6852c390173b26034510d54e6b72296) but kept the comment so it's a no-brain great value work when implementing it (it's going to be in the next few days so! shouldn't stay as it is)

woodenfurniture commented on src/components/ThorFreeFeeBanner/ThorFreeFeeBanner.tsx:
```diff
@@ -0,0 +1,52 @@
+
+export const ThorFreeFeeBanner = () => {
+  return (
+    <Card
+      // as={Link}
+      // href='#'
+      width='full'
+      maxWidth='500px'
+      transition='.25s ease-out'
+      // _hover={cardHover}
+      position='relative'
+      overflow='hidden'
+      mb={4}
+      _before={cardBefore}
+    >
```
nitpick:

```suggestion
    <Card
      width='full'
      maxWidth='500px'
      transition='.25s ease-out'
      position='relative'
      overflow='hidden'
      mb={4}
      _before={cardBefore}
    >
```

NeOMakinG replied:
See comment before

woodenfurniture commented on src/components/ThorFreeFeeBanner/ThorFreeFeeBanner.tsx:
```diff
@@ -0,0 +1,52 @@
+        <Box maxWidth='350px' ml='auto' width='full'>
+          <Text translation='thorFees.title' fontSize='sm' fontWeight='bold' />
+          <Text translation='thorFees.description' my={1} fontSize='sm' color='text.subtle' />
+          {/* <Text translation='common.learnMore' fontSize='sm' fontWeight='bold' color='blue.300' /> */}
```
nitpick:
```suggestion
```

NeOMakinG replied:
See comment before

woodenfurniture commented on src/lib/fees/model.ts:
```diff
@@ -108,6 +124,16 @@ export const calculateFees: CalculateFeeBps = ({ tradeAmountUsd, foxHeld, feeMod
   const feeUsd = feeUsdBeforeDiscount.minus(feeUsdDiscount)
   const foxDiscountUsd = feeUsdBeforeDiscount.times(foxDiscountPercent.div(100))
 
+  console.log({
+    feeBps: feeBps.toFixed(),
+    feeBpsFloat: feeBpsFloat.toFixed(),
+    feeUsd: feeUsd.toFixed(),
+    foxDiscountPercent: foxDiscountPercent.toFixed(),
+    foxDiscountUsd: foxDiscountUsd.toFixed(),
+    feeUsdBeforeDiscount: feeUsdBeforeDiscount.toFixed(),
+    feeBpsBeforeDiscount: feeBpsBeforeDiscount.toFixed(),
+  })
+
```
nitpick:

```suggestion
```

NeOMakinG replied:
fml, changed in [30b7cb9](https://github.com/shapeshift/web/pull/8044/commits/30b7cb926fe202f7e1e9e44fd5a3616439f2f4f9)

woodenfurniture reviewed (changes requested):
Looking very nearly good to go, just a few comments and q's before we get this in!

NeOMakinG commented on src/state/apis/snapshot/snapshot.ts:
```diff
@@ -152,6 +194,63 @@ export const snapshotApi = createApi({
+            }, new Set()),
+          )
+          const delegation = false // don't let people delegate for discounts - ambiguous in spec
+          const votingPowerResults = await Promise.all(
```
We absolutely want a partial success which should already be the case

This is also the case for FOX voting power, should we do both in a follow up or partially do it in this one and include fox in another one?

woodenfurniture replied:
If any of the `getVotingPower` calls throw the `Promise.all` will also throw, so if we want partial failure/success we can use `Promise.allSettled` combined with introspection if thats what we want

NeOMakinG replied:
Yep it would make sense, but I would have to also modify the fox power side which is unrelated to this PR, should I do it in this one or reserve it for a follow up?

NeOMakinG replied:
Changed both in [3ae7cc6](https://github.com/shapeshift/web/pull/8044/commits/3ae7cc687304721da8a65d8b8a8c60ba663fdada)

NeOMakinG commented on src/state/apis/snapshot/snapshot.ts:
```diff
@@ -152,6 +194,63 @@ export const snapshotApi = createApi({
+            return strategiesResult?.data
+          })()
+          if (!strategies) {
+            console.log('snapshotApi getThorVotingPower could not get strategies')
```
Also changed in getVotingPower, in [30b7cb9](https://github.com/shapeshift/web/pull/8044/commits/30b7cb926fe202f7e1e9e44fd5a3616439f2f4f9)

NeOMakinG commented on src/state/apis/snapshot/snapshot.ts:
```diff
@@ -152,6 +194,63 @@ export const snapshotApi = createApi({
+              return acc
+            }, new Set()),
+          )
+          const delegation = false // don't let people delegate for discounts - ambiguous in spec
+          const votingPowerResults = await Promise.all(
+            evmAddresses.map(async address => {
+              const votingPowerUnvalidated = await getVotingPower(
+                address,
+                '1',
+                strategies,
+                THOR_TIP_014_BLOCK_NUMBER,
+                THORSWAP_SNAPSHOT_SPACE,
+                delegation,
+              )
```
Probably never as it's only 2 call then it's even kept in cache somehow as I don't have a new request when I reload the page

We also have a retry, we might want to throttle if we notice any rate limits issues but can we do it in a follow up PR or if we notice an issue?

woodenfurniture replied:
Sweet yeah only 2 addresses should be totally fine. For some reason in my head `evmAddresses` could be like 150 or something but no probs.

NeOMakinG replied:
> Sweet yeah only 2 addresses should be totally fine. For some reason in my head `evmAddresses` could be like 150 or something but no probs.

Ah I think I missed that part, I was thinking about only one call but it effectively do 1 call by address which is already an issue on the fox side

It's obviously an issue as we have 100 request by mins which would fail for 50 EVM addresses, creating an issue for a follow up that I'll do tomorrow if you don't mind (https://github.com/shapeshift/web/issues/8048)

gomesalexandre commented on src/state/apis/snapshot/snapshot.ts:
```diff
@@ -152,6 +194,63 @@ export const snapshotApi = createApi({
+          // Return an error tuple in case of an invalid foxHeld value so we don't cache an errored value
+          if (isNaN(thorHeld)) {
+            const data = 'NaN thorHeld value'
+            return { error: { data, status: 400 } }
```
Totally irrelevant here, we abuse RTK API re: errors but actually never consume this https://redux-toolkit.js.org/rtk-query/usage/error-handling

kaladinlight commented on src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx:
```diff
@@ -211,6 +215,13 @@ export const useTradeExecution = (
+            bnOrZero(thorVotingPower).toNumber() >= THORSWAP_UNIT_THRESHOLD &&
+            new Date().getUTCFullYear() < THORSWAP_MAXIMUM_YEAR_TRESHOLD
+          ) {
+            trackMixpanelEvent(MixPanelEvent.ThorDiscountTradeSuccess)
```
it may be beneficial to track the full state of thor discount trades (confirm, success, and failed)

NeOMakinG replied:
Done in [05daabc](https://github.com/shapeshift/web/pull/8044/commits/05daabc89d2082b788ce937e871349e89d2193ba)

kaladinlight commented on src/lib/fees/parameters/thorchainLp.ts:
```diff
@@ -1,5 +1,6 @@
 import type { FeeCurveParameters } from './types'
 
+// THOR asset doesn't have any fee models for now as it's not using curves, we only zero out values to make TS happy with less risks
```
comment on wrop fee param

NeOMakinG replied:
Done in [05daabc](https://github.com/shapeshift/web/pull/8044/commits/05daabc89d2082b788ce937e871349e89d2193ba)

kaladinlight commented on src/lib/fees/parameters/thor.ts:
```diff
@@ -0,0 +1,19 @@
+import type { FeeCurveParameters } from './types'
+
+const FEE_CURVE_MAX_FEE_BPS = 0 // basis points
+const FEE_CURVE_MIN_FEE_BPS = 0 // basis points
+const FEE_CURVE_NO_FEE_THRESHOLD_USD = 0 // usd
+const FEE_CURVE_FOX_MAX_DISCOUNT_THRESHOLD = 0 // fox
+const FEE_CURVE_MIDPOINT_USD = 0 // usd
+const FEE_CURVE_STEEPNESS_K = 0 // unitless
+const FEE_CURVE_FOX_DISCOUNT_DELAY_HOURS = 0 // 7 for thorchain LP, see https://discord.com/channels/554694662431178782/920449863408287855/1207415430025580627
```
nitpick: these comments are pretty irrelevant and the last one is unrelated if we just want to remove and have the single top level comment

NeOMakinG replied:
Done in [05daabc](https://github.com/shapeshift/web/pull/8044/commits/05daabc89d2082b788ce937e871349e89d2193ba)

kaladinlight commented on src/state/apis/snapshot/snapshot.ts:
```diff
@@ -121,7 +165,7 @@ export const snapshotApi = createApi({
           )
           const delegation = false // don't let people delegate for discounts - ambiguous in spec
-          const votingPowerResults = await Promise.all(
+          const votingPowerResults = await Promise.allSettled(
```
I don't think we want to change this tbh. The idea was that if we can't fetch a users voting power correctly and in a complete manner, that is on us and we default to free trades since we can't make a judgement. This change would potentially result in someone that should have the correct voting power to achieve free trades, to be charged depending on what may or may not have been fetched.

kaladinlight replied:
as per dms, I forgot we changed default to max bps so best effort is actually better for user now. disregard, this makes sense.

kaladinlight commented on src/lib/fees/model.ts:
```diff
@@ -48,15 +53,26 @@ export const calculateFees: CalculateFeeBps = ({ tradeAmountUsd, foxHeld, feeMod
-  const isFree = tradeAmountUsd.lt(noFeeThresholdUsd)
+  const isFree =
+    tradeAmountUsd.lt(noFeeThresholdUsd) ||
+    (isThorFreeEnabled && thorHeld?.isGreaterThanOrEqualTo(THORSWAP_UNIT_THRESHOLD))
```
tracking this in `isThorFree`

NeOMakinG replied:
Done in [05daabc](https://github.com/shapeshift/web/pull/8044/commits/05daabc89d2082b788ce937e871349e89d2193ba)

kaladinlight commented:
making a note for posterity that we are currently displaying the thor discount as a fox power discount. easy mode just disable the modal for thor free trades so we don't have to mess with more UI that will only be there for 2 months.
![image](https://github.com/user-attachments/assets/e97f325b-3ad4-4cfe-9086-bf2e2f8efd31)

kaladinlight commented:
Also noticing what looks like an outstanding bug and potential improvement related to `getVotingPower`.

Bug: switch wallet does not correctly fetch the new/appropriate voting power for the new wallet. i started with my native wallet and then switched to the frame wallet with test address. after switching the wallet, the discount being applied is still for my native wallet.

Improvement: you can start interacting with the trade widget to get quotes immediately and get a quote with fees displayed before the snapshot logic completes. once the snapshot voting power completes, it triggers a refresh of the trade input component and looks weird. seems like we could add a loading state check to keep the quotes displayed, but the fee breakdown loading until voting power fetch is completed and then show (no refresh of state).
